rest_framework-stubs/generics.pyi

```diff
@@ -9,28 +9,30 @@ from rest_framework.request import Request
 from rest_framework.response import Response
 from rest_framework.serializers import BaseSerializer
 
-_T = TypeVar("_T", bound=Model)
+_MT_co = TypeVar("_MT_co", bound=Model, covariant=True)
 
 def get_object_or_404(
-    queryset: Union[Type[_T], Manager[_T], QuerySet[_T]], *filter_args: Any, **filter_kwargs: Any
-) -> _T: ...
+    queryset: Union[Type[_MT_co], Manager[_MT_co], QuerySet[_MT_co]], *filter_args: Any, **filter_kwargs: Any
+) -> _MT_co: ...
 
-class GenericAPIView(views.APIView):
-    queryset: Optional[QuerySet] = ...
+class UsesQuerySet(Protocol[_MT_co]):
+    def get_queryset(self) -> QuerySet[_MT_co]: ...
+
+class GenericAPIView(views.APIView, UsesQuerySet[_MT_co]):
+    queryset: Optional[Union[QuerySet[_MT_co], Manager[_MT_co]]] = ...
     serializer_class: Optional[Type[BaseSerializer]] = ...
     lookup_field: str = ...
     lookup_url_kwarg: Optional[str] = ...
     filter_backends: Sequence[Type[BaseFilterBackend]] = ...
     pagination_class: Optional[Type[BasePagination]] = ...
-    def get_queryset(self) -> QuerySet: ...
-    def get_object(self) -> Any: ...
-    def get_serializer(self, *args: Any, **kwargs: Any) -> BaseSerializer: ...
-    def get_serializer_class(self) -> Type[BaseSerializer]: ...
+    def get_object(self) -> _MT_co: ...
+    def get_serializer(self, *args: Any, **kwargs: Any) -> BaseSerializer[Any, Any, Any, _MT_co]: ...
+    def get_serializer_class(self) -> BaseSerializer[Any, Any, Any, _MT_co]: ...
     def get_serializer_context(self) -> Dict[str, Any]: ...
-    def filter_queryset(self, queryset: QuerySet) -> QuerySet: ...
+    def filter_queryset(self, queryset: QuerySet[_MT_co]) -> QuerySet[_MT_co]: ...
     @property
     def paginator(self) -> Optional[BasePagination]: ...
-    def paginate_queryset(self, queryset: Union[QuerySet, Sequence]) -> Optional[List[Any]]: ...
+    def paginate_queryset(self, queryset: QuerySet[_MT_co]) -> Optional[Sequence[Any]]: ...
```

Comment:
@Goldziher I would say the old typing were better, even if it named paginate *queryset* it actually can work with Sequence too